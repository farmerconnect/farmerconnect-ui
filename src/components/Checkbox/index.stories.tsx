import React from "react";
// also exported from '@storybook/react' if you can deal with breaking changes in 6.1
import { Story, Meta } from "@storybook/react/types-6-0";

import Checkbox, { CheckboxProps } from "./";

export default {
  title: "Checkbox",
  component: Checkbox,
  argTypes: {},
} as Meta;

const Template: Story<CheckboxProps> = (args) => (
  <div>
    <Checkbox {...args} name="radio-0">
      Option 1
    </Checkbox>
    <Checkbox {...args} name="radio-0">
      Option 2
    </Checkbox>
    <Checkbox {...args} name="radio-0">
      Option 3
    </Checkbox>
    <Checkbox {...args} name="radio-0">
      Option 4
    </Checkbox>
    <Checkbox {...args} name="radio-0">
      Option 5
    </Checkbox>
  </div>
);

export const Default = Template.bind({});
Default.args = {};
