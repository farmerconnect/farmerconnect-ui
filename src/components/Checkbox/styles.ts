import styled, { keyframes } from 'styled-components';

const show = keyframes`
0% {
  transform: scale(1.1);
}
100% {
  transform: scale(1);
}
`;

type LabelProps = {
  disabled?: boolean;
};
export const Label = styled.label<LabelProps>`
  box-sizing: border-box;
  color: ${({ disabled }) => (disabled ? '#B9B9B9' : '#141414')};
  font-size: 0.875rem;
  display: inline-flex;
  align-items: center;
  margin-left: 2rem;
  column-gap: 0.25rem;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  position: relative;
  cursor: pointer;

  input {
    cursor: pointer;
    opacity: 0;
    margin: 0;
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
  }

  > svg {
    min-width: 1.5rem;
  }

  input + .unchecked {
    display: block;
  }

  input + .checked {
    display: none;
  }

  input:checked + svg + .unchecked {
    display: none;
  }

  input:checked + .checked {
    display: block;
    animation: ${show} 0.2s;
  }
`;
