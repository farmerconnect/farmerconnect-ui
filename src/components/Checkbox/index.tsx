import { HTMLProps, ReactNode } from 'react';
import { Label } from './styles';
import { CheckboxChecked, CheckboxUnchecked } from '../Icons/Checkbox';

export type CheckboxProps = {
  children?: ReactNode;
} & HTMLProps<HTMLInputElement>;

const Checkbox = ({ children, className, style, ...props }: CheckboxProps) => (
  <Label className={className} style={style}>
    <input type="checkbox" {...props} />
    <CheckboxChecked className="checked" />
    <CheckboxUnchecked data-testid="unchecked" className="unchecked" />
    {children}
  </Label>
);

export default Checkbox;
