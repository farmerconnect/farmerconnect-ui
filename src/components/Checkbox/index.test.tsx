import { render, waitFor, fireEvent } from '@testing-library/react';
import Checkbox from '.';

describe('Checkbox Component', () => {
  it('Renders properly', () => {
    const container = render(
      <div>
        <Checkbox name="radio-test" value={1} placeholder="option-1">
          Option 1
        </Checkbox>
        <Checkbox name="radio-test" value={2} placeholder="option-2">
          Option 2
        </Checkbox>
        <Checkbox name="radio-test" value={3} placeholder="option-3">
          Option 3
        </Checkbox>
      </div>
    );
    expect(container.getByText(/option 1/i)).toBeInTheDocument();
    expect(container.getByText(/option 2/i)).toBeInTheDocument();
    expect(container.getByText(/option 3/i)).toBeInTheDocument();
  });

  it('renders the correct svgs', () => {
    const renderedCheckbox = render(
      <Checkbox name="radio-test" value={1} placeholder="option-1">
        checkbox
      </Checkbox>
    );
    expect(renderedCheckbox.container).toMatchSnapshot();
  });

  it('hides the unchecked svg on click', async () => {
    const renderedCheckbox = render(
      <Checkbox name="radio-test" value={1} placeholder="option-1">
        checkbox
      </Checkbox>
    );

    fireEvent.click(renderedCheckbox.getByLabelText(/checkbox/));
    await waitFor(async () => {
      const element = await renderedCheckbox.findByTestId('unchecked');
      expect(element).toHaveStyle('display: none');
    });
  });
});
