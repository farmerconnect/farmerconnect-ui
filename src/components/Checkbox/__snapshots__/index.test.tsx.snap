// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox Component renders the correct svgs 1`] = `
.c0 {
  box-sizing: border-box;
  color: #141414;
  font-size: 0.875rem;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-left: 2rem;
  -webkit-column-gap: 0.25rem;
  column-gap: 0.25rem;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  position: relative;
  cursor: pointer;
}

.c0 input {
  cursor: pointer;
  opacity: 0;
  margin: 0;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
}

.c0 > svg {
  min-width: 1.5rem;
}

.c0 input + .unchecked {
  display: block;
}

.c0 input + .checked {
  display: none;
}

.c0 input:checked + svg + .unchecked {
  display: none;
}

.c0 input:checked + .checked {
  display: block;
  -webkit-animation: gZMnEj 0.2s;
  animation: gZMnEj 0.2s;
}

<div>
  <label
    class="c0"
  >
    <input
      name="radio-test"
      placeholder="option-1"
      type="checkbox"
      value="1"
    />
    <svg
      class="checked"
      fill="none"
      height="24"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <rect
        fill="#141414"
        height="15"
        rx="3.5"
        stroke="#141414"
        width="15"
        x="4.5"
        y="4.5"
      />
      <path
        d="M10.1348 14.1591L8.03182 12.0561C7.91859 11.9426 7.76486 11.8788 7.60455 11.8788C7.44423 11.8788 7.2905 11.9426 7.17727 12.0561C6.94091 12.2924 6.94091 12.6742 7.17727 12.9106L9.71061 15.4439C9.94697 15.6803 10.3288 15.6803 10.5652 15.4439L16.9773 9.03181C17.2136 8.79545 17.2136 8.41363 16.9773 8.17726C16.864 8.06378 16.7103 8 16.55 8C16.3897 8 16.236 8.06378 16.1227 8.17726L10.1348 14.1591Z"
        fill="white"
      />
    </svg>
    <svg
      class="unchecked"
      data-testid="unchecked"
      fill="none"
      height="24"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <rect
        fill="none"
        height="15"
        rx="3.5"
        stroke="#141414"
        width="15"
        x="4.5"
        y="4.5"
      />
    </svg>
    checkbox
  </label>
</div>
`;
